File: 22.Generate-Parentheses.py
# https://leetcode.com/problems/generate-parentheses/description/
#
# algorithms
# Medium (50.01%)
# Total Accepted:    242.4K
# Total Submissions: 484.7K
# beats 100.0% of python submissions


class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        def resursive(path, left_num, right_num):
            """
            :type path: str
            :type left_num: int
            :type right_num: int
            :rtype: None
            """
            if left_num == 0 and right_num == 0:
                res.append(path)
                return

            if left_num < right_num:
                resursive(path + ')', left_num, right_num - 1)

            if left_num > 0:
                resursive(path + '(', left_num - 1, right_num)

        res = []

        resursive('', n, n)

        return res

File: 22.Generate-Parentheses.java
// https://leetcode.com/problems/generate-parentheses/
//
// algorithms
// Medium (54.86%)
// Total Accepted: 337,678
// Total Submissions: 615,535
// beats 100.0% of java submissions


class Solution {
    public List<String> generateParenthesis(int n) {
        ArrayList<String> res = new ArrayList<>();
        char[] chArr = new char[2 * n];

        recursive(res, n, n, chArr, 0);

        return res;
    }

    public void recursive(ArrayList<String> res, int left, int right, char[] chArr, int idx) {
        if (left == 0 && right == 0) {
            res.add(new String(chArr));
            return;
        }

        if (right > left) {
            chArr[idx] = ')';
            recursive(res, left, right - 1, chArr, idx + 1);
        }

        if (left > 0) {
            chArr[idx] = '(';
            recursive(res, left - 1, right, chArr, idx + 1);
        }
    }
}